src/session.rs

```diff
@@ -337,7 +344,16 @@ where
                         );
                     }
 
-                    Ok(None) => {}
+                    // ...otherwise `store_session` returns `None`.
+                    Ok(None) => {
+                        if let Some(cookie_value) = cookie_value {
+                            let cookie = session_layer.build_cookie(cookie_value);
+                            response.headers_mut().insert(
+                                SET_COOKIE,
+                                HeaderValue::from_str(&cookie.to_string()).unwrap(),
+                            );
+                        }
+                    }
```

Comment:
Sorry for the late response! I've researched this a bit more, and I think this is unnecessary. If `store_session` returns `Ok(None)` we are completely fine and we needn't to set the same cookie value once again for the client. Also I think `uses_valid_session` is not the right behavior. Other than these two the PR seems good to me.

Reply:
Do you know if `uses_valid_session` is modeling the wrong behavior because `MemoryStore` itself is incorrect? One solution could be to use a different store for that test altho it's a little unfortunate that `MemoryStore` itself might be wrong.

Reply:
No, I don't think it's MemoryStore. If you already have a valid cookie and nothing has effectively changed, then we shouldn't send a Set-Cookie header again. We might test that the underlying data is preserved correctly, for example incrementing some counter and see whether it seems right.

Reply:
That makes sense. I updated the test with a counter as you suggested and it seems to be demonstrating what we'd expect.

Reply:
Yes, I thought about something like this. Great!